[SYCL][Doc] Create  sycl_ext_oneapi_queue_status_query.asciidoc (#5860)
intel/llvm: sycl/doc/extensions/proposed/sycl_ext_intel_queue_status_query.asciidoc

// Set the default source code type in this document to C++,
// for syntax highlighting purposes.  This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

== Notice

[%hardbreaks]
Copyright (C) 2022-2022 Intel Corporation.  All rights reserved.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

This extension is written against the SYCL 2020 revision 4 specification.  All
references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

== Status

This is a proposed extension specification, intended to gather community
feedback.  Interfaces defined in this specification may not be implemented yet
or may be in a preliminary state.  The specification itself may also change in
incompatible ways before it is finalized.  *Shipping software products should
not rely on APIs defined in this specification.*

== Overview

> This extension add `::queue::size()`, `sycl::queue::empty()`, `sycl::queue::get_wait_list()` menber functions.

Reviewer: ```suggestion
This extension add `sycl::queue::size()`, `sycl::queue::empty()`, `sycl::queue::get_wait_list()` member functions.
```

> - `sycl::queue::size()` and `ycl::queue::empty` allow user to querry the number of enqueued commands to a queues.

Reviewer: ```suggestion
- `sycl::queue::size()` and `ycl::queue::empty` allow users to query the number of enqueued commands to a queue.
```

- `sycl::queue::get_wait_list()` is the queue equivalent of `event::get_wait_list()` 

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the

> macro `SYCL_EXT_ONEAPI_QUEUE_STATUS_QUERRY` to one of the values defined in the table

Reviewer: ```suggestion
macro `SYCL_EXT_ONEAPI_QUEUE_STATUS_QUERY` to one of the values defined in the table
```

below.  Applications can test for the existence of this macro to determine if
the implementation supports this feature, or applications can test the macro's
value to determine which of the extension's features the implementation
supports.

> _And follow the text with a table like this *unless the extension is
> "experimental"*.  Note that your table may have more than one row if it
> has multiple versions._

Reviewer: ```suggestion

```

[%header,cols="1,5"]
|===
|Value
|Description

|1
|Initial version of this extension.
|===

=== Signature and Semantic 

> This extension adds the following new member function to the SYCL `queue`

Reviewer: ```suggestion
This extension adds the following new member functions to the SYCL `queue` class.
```

```
namespace sycl {
namespace queue {
 bool empty() const;
 size_t size() const;
 std::vector<event> get_wait_list() const;
}}
```

> The semantic of the new function are: 

Reviewer: ```suggestion
The semantics of the new functions are: 
```

```
empty()
   Returns `true` if `size() == 0`.
```

```
size_t size() const

>    Return the number of enqueued `command` in the queue 

Reviewer: ```suggestion
   Returns the number of enqueued `commands` in the queue 
```

   that have not been completed.
   Since the implementation completes commands from the queue 
   asynchronously, the returned value is a snapshot in time,
   and the actual number of uncompleted commands may be different 
   by the time the function returns.
   Note that the behavior of `size()` is associated with the SYCL `queue`
   and not with the underlying native-backend-object.
```

(based on similar HipSYCL extension https://github.com/illuhad/hipSYCL/blob/develop/doc/extensions.md#hipsycl_ext_queue_wait_list) 
```
std::vector<event> get_wait_list() const

>    Return the list of events such that waiting for all returned events

Reviewer: ```suggestion
   Returns the list of events such that waiting for all returned events
```
